File: system/multitier/backend-java/src/test/java/com/mycompany/myshop/backend/core/services/OrderServiceTest.java
package com.mycompany.myshop.backend.core.services;

import com.mycompany.myshop.backend.core.dtos.PlaceOrderRequest;
import com.mycompany.myshop.backend.core.dtos.PlaceOrderResponse;
import com.mycompany.myshop.backend.core.dtos.external.GetPromotionResponse;
import com.mycompany.myshop.backend.core.dtos.external.ProductDetailsResponse;
import com.mycompany.myshop.backend.core.dtos.external.TaxDetailsResponse;
import com.mycompany.myshop.backend.core.entities.Order;
import com.mycompany.myshop.backend.core.entities.OrderStatus;
import com.mycompany.myshop.backend.core.exceptions.NotExistValidationException;
import com.mycompany.myshop.backend.core.exceptions.ValidationException;
import com.mycompany.myshop.backend.core.repositories.OrderRepository;
import com.mycompany.myshop.backend.core.services.external.ClockGateway;
import com.mycompany.myshop.backend.core.services.external.ErpGateway;
import com.mycompany.myshop.backend.core.services.external.TaxGateway;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OrderServiceTest {

    @Mock
    private OrderRepository orderRepository;
    @Mock
    private ErpGateway erpGateway;
    @Mock
    private TaxGateway taxGateway;
    @Mock
    private ClockGateway clockGateway;
    @Mock
    private CouponService couponService;

    @InjectMocks
    private OrderService orderService;

    private static final Instant NORMAL_TIME = Instant.parse("2025-06-15T10:00:00Z");
    private static final Instant DEC_31_YEAR_END_BLACKOUT = Instant.parse("2025-12-31T23:59:00Z");
    private static final Instant DEC_31_CANCEL_BLACKOUT = Instant.parse("2025-12-31T22:15:00Z");

    @Test
    void placeOrderReturnsOrderNumberStartingWithOrd() {
        givenNormalTime();
        givenProductExists("BOOK-123", new BigDecimal("10.00"));
        givenNoPromotion();
        givenNoDiscount();
        givenTaxRate("US", new BigDecimal("0.10"));

        var response = orderService.placeOrder(buildRequest("BOOK-123", 2, "US"));

        var captor = ArgumentCaptor.forClass(Order.class);
        verify(orderRepository).save(captor.capture());
        assertSavedOrder(captor.getValue(), response);
    }

    @Test
    void placeOrderThrowsWhenOrderedOnYearEndBlackout() {
        when(clockGateway.getCurrentTime()).thenReturn(DEC_31_YEAR_END_BLACKOUT);

        var thrown = catchThrowable(() -> orderService.placeOrder(buildRequest("BOOK-123", 1, "US")));

        assertThat(thrown).isInstanceOf(ValidationException.class)
                .hasMessageContaining("December 31");
    }

    @Test
    void placeOrderThrowsWhenSkuUnknown() {
        givenNormalTime();
        when(erpGateway.getProductDetails("UNKNOWN")).thenReturn(Optional.empty());

        var thrown = catchThrowable(() -> orderService.placeOrder(buildRequest("UNKNOWN", 1, "US")));

        assertThat(thrown).isInstanceOf(ValidationException.class);
        assertThat(((ValidationException) thrown).getFieldName()).isEqualTo("sku");
    }

    @Test
    void placeOrderThrowsWhenCountryUnknown() {
        givenNormalTime();
        givenProductExists("BOOK-123", new BigDecimal("10.00"));
        givenNoPromotion();
        givenNoDiscount();
        when(taxGateway.getTaxDetails("XX")).thenReturn(Optional.empty());

        var thrown = catchThrowable(() -> orderService.placeOrder(buildRequest("BOOK-123", 1, "XX")));

        assertThat(thrown).isInstanceOf(ValidationException.class);
        assertThat(((ValidationException) thrown).getFieldName()).isEqualTo("country");
    }

    @Test
    void deliverOrderTransitionsStatusToDelivered() {
        var order = placedOrder("ORD-001");
        when(orderRepository.findByOrderNumber("ORD-001")).thenReturn(Optional.of(order));

        orderService.deliverOrder("ORD-001");

        assertThat(order.getStatus()).isEqualTo(OrderStatus.DELIVERED);
        verify(orderRepository).save(order);
    }

    @Test
    void deliverOrderThrowsWhenOrderNotFound() {
        when(orderRepository.findByOrderNumber("ORD-999")).thenReturn(Optional.empty());

        var thrown = catchThrowable(() -> orderService.deliverOrder("ORD-999"));

        assertThat(thrown).isInstanceOf(NotExistValidationException.class);
    }

    @Test
    void deliverOrderThrowsWhenOrderAlreadyDelivered() {
        var order = placedOrder("ORD-001");
        order.setStatus(OrderStatus.DELIVERED);
        when(orderRepository.findByOrderNumber("ORD-001")).thenReturn(Optional.of(order));

        var thrown = catchThrowable(() -> orderService.deliverOrder("ORD-001"));

        assertThat(thrown).isInstanceOf(ValidationException.class)
                .hasMessageContaining("cannot be delivered");
    }

    @Test
    void cancelOrderTransitionsStatusToCancelled() {
        givenNormalTime();
        var order = placedOrder("ORD-001");
        when(orderRepository.findByOrderNumber("ORD-001")).thenReturn(Optional.of(order));

        orderService.cancelOrder("ORD-001");

        assertThat(order.getStatus()).isEqualTo(OrderStatus.CANCELLED);
        verify(orderRepository).save(order);
    }

    @Test
    void cancelOrderThrowsDuringDecember31CancellationBlackout() {
        when(clockGateway.getCurrentTime()).thenReturn(DEC_31_CANCEL_BLACKOUT);

        var thrown = catchThrowable(() -> orderService.cancelOrder("ORD-001"));

        assertThat(thrown).isInstanceOf(ValidationException.class)
                .hasMessageContaining("December 31");
    }

    @Test
    void cancelOrderThrowsWhenOrderAlreadyCancelled() {
        givenNormalTime();
        var order = placedOrder("ORD-001");
        order.setStatus(OrderStatus.CANCELLED);
        when(orderRepository.findByOrderNumber("ORD-001")).thenReturn(Optional.of(order));

        var thrown = catchThrowable(() -> orderService.cancelOrder("ORD-001"));

        assertThat(thrown).isInstanceOf(ValidationException.class)
                .hasMessageContaining("already been cancelled");
    }

    private void givenNormalTime() {
        when(clockGateway.getCurrentTime()).thenReturn(NORMAL_TIME);
    }

    private void givenProductExists(String sku, BigDecimal price) {
        var product = new ProductDetailsResponse();
        product.setPrice(price);
        when(erpGateway.getProductDetails(sku)).thenReturn(Optional.of(product));
    }

    private void givenNoPromotion() {
        var promotion = new GetPromotionResponse();
        promotion.setPromotionActive(false);
        promotion.setDiscount(BigDecimal.ONE);
        when(erpGateway.getPromotionDetails()).thenReturn(promotion);
    }

    private void givenTaxRate(String country, BigDecimal rate) {
        var taxDetails = new TaxDetailsResponse();
        taxDetails.setTaxRate(rate);
        when(taxGateway.getTaxDetails(country)).thenReturn(Optional.of(taxDetails));
    }

    private void givenNoDiscount() {
        when(couponService.getDiscount(null)).thenReturn(BigDecimal.ZERO);
    }

    private PlaceOrderRequest buildRequest(String sku, int quantity, String country) {
        var request = new PlaceOrderRequest();
        request.setSku(sku);
        request.setQuantity(quantity);
        request.setCountry(country);
        return request;
    }

    private void assertSavedOrder(Order saved, PlaceOrderResponse response) {
        assertThat(saved.getOrderNumber()).startsWith("ORD-").isEqualTo(response.getOrderNumber());
        assertThat(saved.getOrderTimestamp()).isEqualTo(NORMAL_TIME);
        assertThat(saved.getSku()).isEqualTo("BOOK-123");
        assertThat(saved.getQuantity()).isEqualTo(2);
        assertThat(saved.getCountry()).isEqualTo("US");
        assertThat(saved.getUnitPrice()).isEqualByComparingTo("10.00");
        assertThat(saved.getBasePrice()).isEqualByComparingTo("20.00");
        assertThat(saved.getDiscountRate()).isEqualByComparingTo("0");
        assertThat(saved.getDiscountAmount()).isEqualByComparingTo("0");
        assertThat(saved.getSubtotalPrice()).isEqualByComparingTo("20.00");
        assertThat(saved.getTaxRate()).isEqualByComparingTo("0.10");
        assertThat(saved.getTaxAmount()).isEqualByComparingTo("2.00");
        assertThat(saved.getTotalPrice()).isEqualByComparingTo("22.00");
        assertThat(saved.getStatus()).isEqualTo(OrderStatus.PLACED);
        assertThat(saved.getAppliedCouponCode()).isNull();
    }

    private Order placedOrder(String orderNumber) {
        return new Order(
                orderNumber, NORMAL_TIME, "US", "BOOK-123", 1,
                new BigDecimal("10.00"), new BigDecimal("10.00"),
                BigDecimal.ZERO, BigDecimal.ZERO, new BigDecimal("10.00"),
                new BigDecimal("0.10"), new BigDecimal("1.00"), new BigDecimal("11.00"),
                OrderStatus.PLACED, null
        );
    }
}

File: system/multitier/backend-dotnet/Tests/Unit/OrderServiceTest.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Moq;
using MyCompany.MyShop.Backend.Core.Dtos;
using MyCompany.MyShop.Backend.Core.Dtos.External;
using MyCompany.MyShop.Backend.Core.Entities;
using MyCompany.MyShop.Backend.Core.Exceptions;
using MyCompany.MyShop.Backend.Core.Services;
using MyCompany.MyShop.Backend.Core.Services.External;
using MyCompany.MyShop.Backend.Data;
using Xunit;

namespace MyCompany.MyShop.Backend.Tests.Unit;

public class OrderServiceTest : IDisposable
{
    private static readonly DateTime NormalTime = new(2025, 6, 15, 10, 0, 0, DateTimeKind.Utc);
    private static readonly DateTime Dec31YearEndBlackout = new(2025, 12, 31, 23, 59, 0, DateTimeKind.Utc);
    private static readonly DateTime Dec31CancelBlackout = new(2025, 12, 31, 22, 15, 0, DateTimeKind.Utc);

    private readonly AppDbContext _dbContext;
    private readonly Mock<ClockGateway> _clockMock;
    private readonly Mock<ErpGateway> _erpMock;
    private readonly Mock<TaxGateway> _taxMock;
    private readonly Mock<CouponService> _couponMock;
    private readonly OrderService _service;

    public OrderServiceTest()
    {
        var config = new Mock<IConfiguration>();
        var httpClient = new HttpClient();

        _dbContext = new AppDbContext(new DbContextOptionsBuilder<AppDbContext>()
            .UseInMemoryDatabase($"test-{Guid.NewGuid()}")
            .Options);

        _clockMock = new Mock<ClockGateway>(httpClient, config.Object);
        _erpMock = new Mock<ErpGateway>(httpClient, config.Object);
        _taxMock = new Mock<TaxGateway>(httpClient, config.Object);
        _couponMock = new Mock<CouponService>(_dbContext, _clockMock.Object);

        _service = new OrderService(_dbContext, _erpMock.Object, _taxMock.Object, _clockMock.Object, _couponMock.Object);
    }

    public void Dispose()
    {
        _dbContext.Dispose();
        GC.SuppressFinalize(this);
    }

    [Fact]
    public async Task PlaceOrder_ReturnsOrderNumberStartingWithOrd()
    {
        GivenNormalTime();
        GivenProductExists("BOOK-123", 10.00m);
        GivenNoPromotion();
        GivenNoDiscount();
        GivenTaxRate("US", 0.10m);

        var response = await _service.PlaceOrderAsync(BuildRequest("BOOK-123", 2, "US"));

        Assert.StartsWith("ORD-", response.OrderNumber);
        await AssertSavedOrder(response);
    }

    [Fact]
    public async Task PlaceOrder_ThrowsWhenOrderedOnYearEndBlackout()
    {
        _clockMock.Setup(g => g.GetCurrentTimeAsync()).ReturnsAsync(Dec31YearEndBlackout);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.PlaceOrderAsync(BuildRequest("BOOK-123", 1, "US")));

        Assert.Contains("December 31", ex.Message);
    }

    [Fact]
    public async Task PlaceOrder_ThrowsWhenSkuUnknown()
    {
        GivenNormalTime();
        _erpMock.Setup(g => g.GetProductDetailsAsync("UNKNOWN")).ReturnsAsync((ProductDetailsResponse?)null);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.PlaceOrderAsync(BuildRequest("UNKNOWN", 1, "US")));

        Assert.Equal("sku", ex.FieldName);
    }

    [Fact]
    public async Task PlaceOrder_ThrowsWhenCountryUnknown()
    {
        GivenNormalTime();
        GivenProductExists("BOOK-123", 10.00m);
        GivenNoPromotion();
        GivenNoDiscount();
        _taxMock.Setup(g => g.GetTaxDetailsAsync("XX")).ReturnsAsync((TaxDetailsResponse?)null);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.PlaceOrderAsync(BuildRequest("BOOK-123", 1, "XX")));

        Assert.Equal("country", ex.FieldName);
    }

    [Fact]
    public async Task DeliverOrder_TransitionsStatusToDelivered()
    {
        await SeedOrder("ORD-001", OrderStatus.PLACED);

        await _service.DeliverOrderAsync("ORD-001");

        var saved = await _dbContext.Orders.FirstAsync(o => o.OrderNumber == "ORD-001");
        Assert.Equal(OrderStatus.DELIVERED, saved.Status);
    }

    [Fact]
    public async Task DeliverOrder_ThrowsWhenOrderNotFound()
    {
        await Assert.ThrowsAsync<NotExistValidationException>(
            () => _service.DeliverOrderAsync("ORD-999"));
    }

    [Fact]
    public async Task DeliverOrder_ThrowsWhenOrderAlreadyDelivered()
    {
        await SeedOrder("ORD-001", OrderStatus.DELIVERED);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.DeliverOrderAsync("ORD-001"));

        Assert.Contains("cannot be delivered", ex.Message);
    }

    [Fact]
    public async Task CancelOrder_TransitionsStatusToCancelled()
    {
        GivenNormalTime();
        await SeedOrder("ORD-001", OrderStatus.PLACED);

        await _service.CancelOrderAsync("ORD-001");

        var saved = await _dbContext.Orders.FirstAsync(o => o.OrderNumber == "ORD-001");
        Assert.Equal(OrderStatus.CANCELLED, saved.Status);
    }

    [Fact]
    public async Task CancelOrder_ThrowsDuringDecember31CancellationBlackout()
    {
        _clockMock.Setup(g => g.GetCurrentTimeAsync()).ReturnsAsync(Dec31CancelBlackout);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.CancelOrderAsync("ORD-001"));

        Assert.Contains("December 31", ex.Message);
    }

    [Fact]
    public async Task CancelOrder_ThrowsWhenOrderAlreadyCancelled()
    {
        GivenNormalTime();
        await SeedOrder("ORD-001", OrderStatus.CANCELLED);

        var ex = await Assert.ThrowsAsync<ValidationException>(
            () => _service.CancelOrderAsync("ORD-001"));

        Assert.Contains("already been cancelled", ex.Message);
    }

    private void GivenNormalTime() =>
        _clockMock.Setup(g => g.GetCurrentTimeAsync()).ReturnsAsync(NormalTime);

    private void GivenProductExists(string sku, decimal price) =>
        _erpMock.Setup(g => g.GetProductDetailsAsync(sku))
            .ReturnsAsync(new ProductDetailsResponse { Price = price });

    private void GivenNoPromotion() =>
        _erpMock.Setup(g => g.GetPromotionDetailsAsync())
            .ReturnsAsync(new GetPromotionResponse { PromotionActive = false, Discount = 1m });

    private void GivenNoDiscount() =>
        _couponMock.Setup(s => s.GetDiscountAsync(null)).ReturnsAsync(0m);

    private void GivenTaxRate(string country, decimal rate) =>
        _taxMock.Setup(g => g.GetTaxDetailsAsync(country))
            .ReturnsAsync(new TaxDetailsResponse { TaxRate = rate });

    private static PlaceOrderRequest BuildRequest(string sku, int quantity, string country) =>
        new() { Sku = sku, Quantity = quantity, Country = country };

    private async Task SeedOrder(string orderNumber, OrderStatus status)
    {
        _dbContext.Orders.Add(new Order
        {
            OrderNumber = orderNumber,
            OrderTimestamp = NormalTime,
            Country = "US",
            Sku = "BOOK-123",
            Quantity = 1,
            UnitPrice = 10m,
            BasePrice = 10m,
            DiscountRate = 0m,
            DiscountAmount = 0m,
            SubtotalPrice = 10m,
            TaxRate = 0.10m,
            TaxAmount = 1m,
            TotalPrice = 11m,
            Status = status
        });
        await _dbContext.SaveChangesAsync();
    }

    private async Task AssertSavedOrder(PlaceOrderResponse response)
    {
        var saved = await _dbContext.Orders.FirstAsync();
        Assert.StartsWith("ORD-", saved.OrderNumber);
        Assert.Equal(response.OrderNumber, saved.OrderNumber);
        Assert.Equal(NormalTime, saved.OrderTimestamp);
        Assert.Equal("BOOK-123", saved.Sku);
        Assert.Equal(2, saved.Quantity);
        Assert.Equal("US", saved.Country);
        Assert.Equal(10.00m, saved.UnitPrice);
        Assert.Equal(20.00m, saved.BasePrice);
        Assert.Equal(0m, saved.DiscountRate);
        Assert.Equal(0m, saved.DiscountAmount);
        Assert.Equal(20.00m, saved.SubtotalPrice);
        Assert.Equal(0.10m, saved.TaxRate);
        Assert.Equal(2.00m, saved.TaxAmount);
        Assert.Equal(22.00m, saved.TotalPrice);
        Assert.Equal(OrderStatus.PLACED, saved.Status);
        Assert.Null(saved.AppliedCouponCode);
    }
}
